File: booking-api/services/calendar.py
"""Google Calendar service — creates calendar events for confirmed bookings."""

import json
import os
from datetime import datetime

from google.oauth2 import service_account
from googleapiclient.discovery import build

from config import settings


def _get_calendar_service():
    """Build and return an authenticated Google Calendar API service."""
    creds = None

    if settings.google_service_account_json:
        info = json.loads(settings.google_service_account_json)
        creds = service_account.Credentials.from_service_account_info(
            info,
            scopes=["https://www.googleapis.com/auth/calendar"],
        )
    elif settings.google_service_account_file and os.path.exists(settings.google_service_account_file):
        creds = service_account.Credentials.from_service_account_file(
            settings.google_service_account_file,
            scopes=["https://www.googleapis.com/auth/calendar"],
        )
    else:
        raise RuntimeError(
            "No Google service account credentials found. "
            "Set GOOGLE_SERVICE_ACCOUNT_FILE or GOOGLE_SERVICE_ACCOUNT_JSON in .env"
        )

    return build("calendar", "v3", credentials=creds)


def create_booking_event(booking: dict) -> dict | None:
    """Create a Google Calendar event for a confirmed booking.

    Returns the created event dict, or None on failure.
    """
    if not settings.google_sheets_spreadsheet_id:
        print("⚠️ GOOGLE_SHEETS_SPREADSHEET_ID not set — skipping calendar event")
        return None

    service = _get_calendar_service()
    calendar_id = settings.google_calendar_id or "primary"

    preferred_date = booking.get("preferred_date", "")
    start_time = booking.get("preferred_time", booking.get("start_time", "10:00"))

    # Use end_time from booking if provided, otherwise calculate (default 1 hour)
    end_time_from_booking = booking.get("end_time", "")
    if end_time_from_booking:
        end_time = end_time_from_booking
    else:
        start_h, start_m = map(int, start_time.split(":"))
        end_h = start_h + int(settings.slot_duration_hours)
        end_m = start_m
        if end_h >= 24:
            end_h = 23
            end_m = 59
        end_time = f"{end_h:02d}:{end_m:02d}"

    # Format as ISO 8601 with IST offset
    start_iso = f"{preferred_date}T{start_time}:00+05:30"
    end_iso = f"{preferred_date}T{end_time}:00+05:30"

    # Build description
    description_parts = [
        f"Booking ID: {booking.get('booking_id', 'N/A')}",
        f"Customer: {booking.get('client_name', 'N/A')}",
        f"Phone: {booking.get('client_phone', 'N/A')}",
        f"Email: {booking.get('client_email', 'N/A')}",
        f"Service: {booking.get('service_type', 'N/A')}",
        f"Venue: {booking.get('venue_name', 'N/A')}",
        f"Payment: {booking.get('payment_status', 'N/A')}",
        f"Message: {booking.get('message', 'N/A')}",
    ]

    event = {
        "summary": f"{booking.get('service_type', 'Booking')} — {booking.get('client_name', '')}",
        "description": "\n".join(description_parts),
        "start": {
            "dateTime": start_iso,
            "timeZone": "Asia/Kolkata",
        },
        "end": {
            "dateTime": end_iso,
            "timeZone": "Asia/Kolkata",
        },
        "reminders": {
            "useDefault": False,
            "overrides": [
                {"method": "email", "minutes": 60},
                {"method": "popup", "minutes": 30},
            ],
        },
    }

    try:
        created = service.events().insert(
            calendarId=calendar_id,
            body=event,
            sendUpdates="none",
        ).execute()
        print(f"✅ Calendar event created: {created.get('htmlLink', created.get('id', ''))}")
        return created
    except Exception as e:
        error_str = str(e)
        print(f"❌ Calendar event creation failed: {error_str[:200]}")
        return None

File: netlify/functions/lib/calendar.js
/**
 * Google Calendar service — creates calendar events for confirmed bookings.
 * Ported from booking-api/services/calendar.py
 */

const { GoogleAuth } = require('google-auth-library');
const { calendar: calendarFactory } = require('@googleapis/calendar');

function getAuth() {
  const jsonStr = process.env.GOOGLE_SERVICE_ACCOUNT_JSON;
  if (!jsonStr) {
    throw new Error('GOOGLE_SERVICE_ACCOUNT_JSON not set');
  }
  const credentials = JSON.parse(jsonStr);
  return new GoogleAuth({
    credentials,
    scopes: ['https://www.googleapis.com/auth/calendar'],
  });
}

/** @type {import('@googleapis/calendar').calendar_v3.Calendar | null} */
let _calendarClient = null;

async function getCalendarService() {
  if (_calendarClient) return _calendarClient;
  const auth = getAuth();
  _calendarClient = calendarFactory({ version: 'v3', auth });
  return _calendarClient;
}

function getSlotDurationHours() {
  return parseFloat(process.env.SLOT_DURATION_HOURS || '1');
}

/**
 * Check if the booking is for a past date — skip Calendar event if so.
 */
function isPastBooking(preferredDate) {
  if (!preferredDate) return false;
  const now = new Date();
  const istOffset = 5.5 * 60 * 60 * 1000;
  const istNow = new Date(now.getTime() + istOffset);
  const todayStr = istNow.toISOString().split('T')[0];
  return preferredDate < todayStr;
}

/**
 * Create a Google Calendar event for a confirmed booking.
 * Returns the created event object, or null on failure.
 */
async function createBookingEvent(booking) {
  // Only create Calendar events for future or today's bookings
  if (isPastBooking(booking.preferred_date)) {
    console.log(`⏭️ Skipping Calendar event for past booking ${booking.booking_id} (${booking.preferred_date})`);
    return null;
  }

  if (!process.env.GOOGLE_SERVICE_ACCOUNT_JSON) {
    console.log('⚠️ GOOGLE_SERVICE_ACCOUNT_JSON not set — skipping calendar event');
    return null;
  }

  const service = await getCalendarService();
  const calendarId = process.env.GOOGLE_CALENDAR_ID || 'primary';

  const preferredDate = booking.preferred_date || '';
  const startTime = booking.preferred_time || booking.start_time || '10:00';

  // Use end_time from booking if provided, otherwise calculate
  let endTime;
  if (booking.end_time) {
    endTime = booking.end_time;
  } else {
    const [sh, sm] = startTime.split(':').map(Number);
    const durationH = getSlotDurationHours();
    let eh = sh + Math.floor(durationH);
    let em = sm + Math.round((durationH % 1) * 60);
    if (eh >= 24) { eh = 23; em = 59; }
    endTime = `${String(eh).padStart(2, '0')}:${String(em).padStart(2, '0')}`;
  }

  // Format as ISO 8601 with IST offset
  const startIso = `${preferredDate}T${startTime}:00+05:30`;
  const endIso = `${preferredDate}T${endTime}:00+05:30`;

  // Build description
  const descriptionParts = [
    `Booking ID: ${booking.booking_id || 'N/A'}`,
    `Customer: ${booking.client_name || 'N/A'}`,
    `Phone: ${booking.client_phone || 'N/A'}`,
    `Email: ${booking.client_email || 'N/A'}`,
    `Service: ${booking.service_type || 'N/A'}`,
    `Venue: ${booking.venue_name || 'N/A'}`,
    `Payment: ${booking.payment_status || 'N/A'}`,
    `Amount: ₹${booking.total_amount || 'N/A'}`,
    `Message: ${booking.message || 'N/A'}`,
  ];

  const event = {
    summary: `${booking.service_type || 'Booking'} — ${booking.client_name || ''}`,
    description: descriptionParts.join('\n'),
    start: {
      dateTime: startIso,
      timeZone: 'Asia/Kolkata',
    },
    end: {
      dateTime: endIso,
      timeZone: 'Asia/Kolkata',
    },
    reminders: {
      useDefault: false,
      overrides: [
        { method: 'email', minutes: 60 },
        { method: 'popup', minutes: 30 },
      ],
    },
  };

  try {
    const created = await service.events.insert({
      calendarId,
      requestBody: event,
      sendUpdates: 'none',
    });
    console.log(`✅ Calendar event created: ${created.data.htmlLink || created.data.id || ''}`);
    return created.data;
  } catch (err) {
    console.error(`❌ Calendar event creation failed: ${err.message}`);
    return null;
  }
}

/**
 * List calendar events from a date range.
 * Returns events created by our service account or matching our naming pattern.
 */
async function listCalendarEvents(options = {}) {
  const {
    calendarId = process.env.GOOGLE_CALENDAR_ID || 'primary',
    timeMin = null,
    timeMax = null,
  } = options;

  if (!process.env.GOOGLE_SERVICE_ACCOUNT_JSON) {
    console.log('⚠️ GOOGLE_SERVICE_ACCOUNT_JSON not set — cannot list calendar events');
    return [];
  }

  const service = await getCalendarService();

  const params = {
    calendarId,
    singleEvents: true,
    orderBy: 'startTime',
  };
  if (timeMin) params.timeMin = timeMin;
  if (timeMax) params.timeMax = timeMax;

  try {
    const response = await service.events.list(params);
    const events = response.data.items || [];
    console.log(`📅 Found ${events.length} events in calendar`);
    return events;
  } catch (err) {
    console.error(`❌ Failed to list calendar events: ${err.message}`);
    return [];
  }
}

/**
 * Delete a calendar event by its ID.
 */
async function deleteCalendarEvent(eventId, calendarId = process.env.GOOGLE_CALENDAR_ID || 'primary') {
  if (!process.env.GOOGLE_SERVICE_ACCOUNT_JSON) return false;

  const service = await getCalendarService();

  try {
    await service.events.delete({ calendarId, eventId });
    console.log(`🗑️ Deleted calendar event: ${eventId}`);
    return true;
  } catch (err) {
    console.error(`❌ Failed to delete calendar event ${eventId}: ${err.message}`);
    return false;
  }
}

module.exports = { createBookingEvent, listCalendarEvents, deleteCalendarEvent };
